### DISCUSSION TITLE
How to project an indirect enum value?

### DISCUSSION THREAD

**Discussion by stephencelis** [Azoy/Echo#56] [Q&A]
A previous discussion mentions `projectValue()` as a means of unpacking an enum case's associated values, but this doesn't seem to work for indirect/recursive enum cases.

Given the following enum:

```swift
enum Foo: Equatable {
  indirect case foo(Foo)
  case bar(Int)
}
```

It's easy enough to extract an `Int` from the `bar` case:

```swift
let val = Foo.bar(42)
var box = container(for: val)
let expected = 42
let actual = box.projectValue().load(as: Int.self)
XCTAssertEqual(actual, expected)
```

But attempting to extract a `Foo` from the `foo` case fails (because it returns identity):

```swift
let val = Foo.foo(.bar(42))
var box = container(for: val)
let expected = Foo.bar(42)
let actual = box.projectValue().load(as: Foo.self)
XCTAssertEqual(actual, expected)
// ❌ XCTAssertEqual failed: ("foo(.bar(42))") is not equal to ("bar(42)")
```

Is there a way to properly extract such a case?

**Comment by iampatbrown** [comment]
@stephencelis How'd you end up going with this? I'm keen to get something working. The below works on your example but there's probably a much better way to go about it. Was hoping you could point me in the right direction. 

```swift
extension AnyExistentialContainer {
  public mutating func projectEnumValue() -> UnsafeRawPointer {
    let projectedValue = self.projectValue()
    guard let enumMetadata = metadata as? EnumMetadata else { return projectedValue }
    let tag = enumMetadata.enumVwt.getEnumTag(for: projectedValue)
    let field = enumMetadata.descriptor.fields.records[Int(tag)]
    guard field.flags.isIndirectCase else { return projectedValue }
    let alignMask = UInt(metadata.vwt.flags.alignmentMask)
    let heapObjSize = UInt(MemoryLayout<HeapObject>.size)
    let byteOffset = (heapObjSize + alignMask) & ~alignMask
    let bytePtr = withUnsafePointer(to: &self) {
      $0.withMemoryRebound(to: UnsafePointer<HeapObject>.self, capacity: 1) {
        $0.pointee.raw
      }
    }
    return bytePtr + Int(byteOffset)
  }
}
```

**Reply by stephencelis** [reply]
There is some information in [this Swift blog post](https://swift.org/blog/how-mirror-works/) about how the Mirror API is implemented, and [this implementation using Echo](https://github.com/Azoy/Echo/blob/f3ba95c4f9eeb72a63c84dc6f944fe1852990df9/Sources/EchoMirror/ReflectionMirror.swift). We tried to port the logic here, but were unable to get something that works robustly, and we were beginning to use a lot more of the runtime than feels necessary:

https://github.com/pointfreeco/swift-case-paths/blob/d1af52f98304706156351bb3052e556c871051a2/Sources/CasePaths/EnumReflection.swift#L83

**Reply by iampatbrown** [reply]
Thats awesome :) Didn't think to look at the other branches :P Yeah it started to feel like reimplementing a trimmed down version just for CasePaths would be tough. Thanks for letting me know.
